source/material_model/rheology/strain_dependent.cc

```diff
@@ -415,7 +494,15 @@ namespace aspect
         if  (this->simulator_is_past_initialization() && this->get_timestep_number() > 0 && in.requests_property(MaterialProperties::reaction_terms))
           {
             const double edot_ii = std::max(sqrt(std::fabs(second_invariant(deviator(in.strain_rate[i])))),min_strain_rate);
-            const double e_ii = edot_ii*this->get_timestep();
+            double e_ii = edot_ii*this->get_timestep();
+
+            // Adjusting strain values to account for strain healing without exceeding an unreasonable range
+            if (healing_mechanism != no_healing)
+              {
+                // Never heal more strain than exists
+                const double healed_strain = std::min(calculate_strain_healing(in,i), e_ii);
+                e_ii -= healed_strain;
+              }
             if (weakening_mechanism == plastic_weakening_with_plastic_strain_only && plastic_yielding == true)
               out.reaction_terms[i][this->introspection().compositional_index_for_name("plastic_strain")] = e_ii;
```

Comment:
Each of the lines down here would then need to be changed the following way:

```suggestion
              out.reaction_terms[i][this->introspection().compositional_index_for_name("plastic_strain")] = 
              std::max(delta_e_ii, -in.composition[i][this->introspection().compositional_index_for_name("plastic_strain")]);
```